plugins/samba.lua

```diff
@@ -0,0 +1,130 @@
+if lldap == nil then
+    lldap = {}
+end
+
+local ensure_user_attribute_exists = function(context, schema, attribute_name, attribute_type)
+    if schema.user_attributes.attributes[attribute_name] == nil then
+        local res = context.api:add_user_attribute({
+            name = attribute_name,
+            attribute_type = attribute_type,
+            is_list = false,
+            is_visible = true,
+            is_editable = false,
+        })
+        if res ~= nil then
+            -- Error.
+            error("Got error from creating '" .. attribute_name .. "' attribute", 1)
+        end
+    end
+end
+
+local ensure_group_attribute_exists = function(context, schema, attribute_name, attribute_type)
+    if schema.group_attributes.attributes[attribute_name] == nil then
+        local res = context.api:add_group_attribute({
+            name = attribute_name,
+            attribute_type = attribute_type,
+            is_list = false,
+            is_visible = true,
+            is_editable = false,
+        })
+        if res ~= nil then
+            -- Error.
+            error("Got error from creating '" .. attribute_name .. "' group attribute", 1)
+        end
+    end
+end
+
+local initialize_attributes = function(context)
+    lldap.log:debug("Initializing samba.lua")
+    local schema = context.api:get_schema()
+    -- ensure we have the basic samba related attributes for users
+    ensure_user_attribute_exists(context, schema, "sambantpassword", "String")
+    ensure_user_attribute_exists(context, schema, "sambasid", "String")
+    ensure_user_attribute_exists(context, schema, "sambaacctflags", "String")
+    -- ensure we have the basic samba related attributes for groups
+    ensure_group_attribute_exists(context, schema, "sambasid", "String")
+    -- ensure we have the 'sambaSamAccount' user object class
+    if schema.extra_user_object_classes.sambaSamAccount == nil then
+        local res = context.api:add_user_object_class("sambaSamAccount")
+        if res ~= nil then
+            -- Error.
+            lldap.log:warn("Got error from creating 'sambaSamAccount' user object class")
+            return res
+        end
+    end
+end
+
+local on_password_update = function(context, args)
+    lldap.log:debug("New password being set for " .. args.user_id)
+    utf16_bytes = lldap.strings:to_utf16le(args.password)
+    hashed = lldap.hashing:md4_hash_bytes(utf16bytes)
```

Comment:
I still think we could give the md4 of the (utf-16) password instead of the cleartext

Reply:
I like the permission model idea, where you can choose a number of different types of hashed passwords, yes, but I also think it's convenient to allow one of the options to be cleartext. Maybe the plugin handles keeping some external, non-LDAP compatible system in sync with password -- or need it for implementing some alternative bind auth mechanism.

An MD4'ed UTF16LE value would be fine for my use, though, and what I'd use.

Reply:
Let's start with just different flavors of hashing, leave room in the design for plaintext, and let the potential users come and argue for it.

Reply:
Implemented a version of the solution discussed on Discord. Avoids passing a cleartext password.

Marking this as resolved; there'll be review of the implemented solution later.